### PR TITLE
feat(ui): Replace GPL compatibility by OSI approved/FSF libre in License

### PR DESCRIPTION
**Pull request by quan-tv**
Signed-off-by: Tran Vu Quan <quan1.tranvu@toshiba.co.jp>

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> - Replace GPL compatibility by OSI approved and FSF libre in License object
> - Update view/add/edit license page
> - Update Export Spreadsheet function
> - Update Import SPDX Information function to get OSI approved and FSF libre information from SPDX

Issue: #1166

### Checklist
Must:
- [x] All related issues are referenced in commit messages and in PR

### CODE REVIEW

File: libraries/lib-datahandler/src/main/thrift/sw360.thrift [hunk 1]
```diff
@@ -12,6 +12,11 @@
 namespace java org.eclipse.sw360.datahandler.thrift
 namespace php sw360.thrift
 
+enum Quadratic {
```
Comment (akapti) [inline libraries/lib-datahandler/src/main/thrift/sw360.thrift:15]: Can we re-use the existing `Ternary` enum? or can you me  let's know the specific need for this new `Quadratic` enum?

**Review by akapti — CHANGES_REQUESTED** [review]
reviewed the code, suggested minor change.
tested the changes, and it's working fine.

File: libraries/lib-datahandler/src/main/thrift/sw360.thrift (hunk 1 above)
Comment (quan-tv) [reply libraries/lib-datahandler/src/main/thrift/sw360.thrift:15]: `Ternary` got 3 values while we only need 2 values for theses fields. So, we decided to create new enum.

File: libraries/exporters/src/main/java/org/eclipse/sw360/exporter/utils/ConvertRecord.java [hunk 2]
```diff
@@ -318,13 +318,13 @@ public static void fillTodoCustomPropertyInfo(List<Obligation> obligations, List
...
-                    final ArrayList<String> out = new ArrayList<>(8);
+                    final ArrayList<String> out = new ArrayList<>(10);
                     out.add(CommonUtils.nullToEmptyString(license.getId()));
                     out.add(CommonUtils.nullToEmptyString(license.getFullname()));
                     out.add(license.isSetLicenseType() ? ((Integer) license.getLicenseType().getLicenseTypeId()).toString() :
                             CommonUtils.nullToEmptyString(license.getLicenseTypeDatabaseId()));
-                    out.add(license.isSetGPLv2Compat() ? (license.getGPLv2Compat()).toString() : "");
-                    out.add(license.isSetGPLv3Compat() ? (license.getGPLv3Compat()).toString() : "");
+		    out.add(license.isSetOSIApproved() ? (license.getOSIApproved()).toString() : "");
+		    out.add(license.isSetFSFLibre() ? (license.getFSFLibre()).toString() : "");
```
[the remark this answers is not in the document]
Comment (quan-tv) [reply libraries/exporters/src/main/java/org/eclipse/sw360/exporter/utils/ConvertRecord.java]: I updated source code follow your comment.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/tags/DisplayBoolean.java [hunk 3]
```diff
@@ -52,9 +57,10 @@ public void doTag() throws JspException, IOException {
             out.print("&nbsp;"+LanguageUtil.get(resourceBundle,"yes"));
             out.print("</span>");
         } else {
+            String text = (name != null && name.equals("Quadratic")) ? "(n/a)" : "no";
```
[the remark this answers is not in the document]
Comment (quan-tv) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/tags/DisplayBoolean.java]: I updated source code follow your comment.

**Review by akapti — APPROVED** [review]
Code looks good.

File: libraries/lib-datahandler/src/main/thrift/licenses.thrift [hunk 4]
```diff
@@ -84,12 +84,10 @@ struct License {
     300: optional map<string, string> additionalData,
 
     // Additional informations
-	// 10: optional bool GPLv2Compat,
-	// 11: optional bool GPLv3Compat,
 	12: optional string reviewdate,
 
-	15: optional Ternary GPLv2Compat = sw360.Ternary.UNDEFINED,
```
Comment (mcjaeger) [inline libraries/lib-datahandler/src/main/thrift/licenses.thrift:91]: pls. let's check if a migration would be required and also ... it is recommended to use different numbers (for new fields) even if already used are not used anymore. so proposal would be to use 17 and 18 as numbering.

Comment (quan-tv) [reply libraries/lib-datahandler/src/main/thrift/licenses.thrift:91]: Pardon for my dummy question, about 'migration', do you mean that I should prepare script for migration similar to theses script:

https://github.com/eclipse/sw360/tree/master/scripts/migrations

Comment (quan-tv) [reply libraries/lib-datahandler/src/main/thrift/licenses.thrift:91]: I've pushed new commit:
- Refine enum numbering
- Add migration script

Could you please review?

File: scripts/migrations/043_replace_gpl_compatibility_by_osi_approved.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python
+# -----------------------------------------------------------------------------
+# Copyright Toshiba corporation, 2021. Part of the SW360 Portal Project.
+# Copyright Toshiba Software Development (Vietnam) Co., Ltd., 2021. Part of the SW360 Portal Project.
+# Part of the SW360 Portal Project.
```
Comment (KoukiHama) [inline scripts/migrations/043_replace_gpl_compatibility_by_osi_approved.py:5]: L5  is Redundant.

Comment (quan-tv) [reply scripts/migrations/043_replace_gpl_compatibility_by_osi_approved.py]: I removed this line.

**Comment by KoukiHama** [comment]
Can someone please review our code again?

**Comment by akapti** [comment]
@quan-tv Can you please resolve the merge conflicts?

**Comment by quan-tv** [comment]
@akapti 
Conflicts were resolved!

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/LicenseSpecTest.java [hunk 6]
```diff
@@ -146,6 +146,8 @@ public void should_document_get_license() throws Exception {
                                 fieldWithPath("additionalData").description("A place to store additional data used by external tools"),
                                 fieldWithPath("text").description("The license's original text"),
                                 fieldWithPath("checked").description("The information, whether the license is already checked, optional and defaults to true"),
+                                fieldWithPath("OSIApproved").description("The OSI aprroved information"),
+                                fieldWithPath("FSFLibre").description("The FSF libre information"),
```
Comment (akapti) [inline rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/LicenseSpecTest.java:150]: Please change the description to add the possible values as shown below.:
`fieldWithPath("OSIApproved").description("The OSI aprroval information, possible values are: " + Arrays.asList(Ternary.values())),`

`fieldWithPath("FSFLibre").description("The FSF libre information, possible values are: " + Arrays.asList(Ternary.values())),`

**Review by akapti — CHANGES_REQUESTED** [review]
Tested the PR, found an issue in REST endpoint  for **Listing licenses**, 
Even if the value for `OSIApproved` and `FSFLibre` is set to `YES`, the REST response for `http://localhost:8080/resource/api/licenses/` always returns value as `NA`.

Tested the migration script, as is working fine. Kindly rename script to `044_replace_gpl_compatibility_by_osi_approved.py`, as there is already another script with `043`. and also provide same read/write/execute access for newly added migration script.

image

**Comment by quan-tv** [comment]
@akapti 
I added new commit, update the source code follow your comments:
- Fix the side-effect of License API
- Update API document
- Rename migration script file

Could you please re-check>

**Review by akapti — COMMENTED** [review]
Code looks good.
Code changes and migration script working as expected.

Kindly change permission of script from `-rw-rw-r--` to `-rwxr-xr-x` as it's there for other migration script.
image

**Comment by quan-tv** [comment]
@akapti 
I've changed the file permission to 755 (`-rwxr-xr-x`), but in my environment, all other files got the permission 775 (`-rwxrwxr-x`).

image

Could you please confirm that it is normal phenomenon?

**Review by akapti — APPROVED** [review]

**Pull request merged**